hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -1498,13 +1479,18 @@ def draw_integer(
         if forced is not None and max_value is not None:
             assert forced <= max_value
 
-        return self.provider.draw_integer(
-            min_value=min_value,
-            max_value=max_value,
-            weights=weights,
-            shrink_towards=shrink_towards,
-            forced=forced,
-        )
+        kwargs = {
+            "min_value": min_value,
+            "max_value": max_value,
+            "weights": weights,
+            "shrink_towards": shrink_towards,
+        }
+        self.start_example(DRAW_INTEGER_LABEL)
+        value = self.provider.draw_integer(**kwargs, forced=forced)
+        self.stop_example()
+        if observe:
+            self.observer.draw_integer(value, forced=forced is not None, kwargs=kwargs)
+        return value
```

Comment:
There are two types of draws: sub-ir and super-ir. We don't want to write the former to the tree, since they are fully recorded by writing the ir node itself. But we do want to write any super ir draws to the tree.

`observe` formalizes/implements this split. `observe=False` for the former, which will skip `self.observer.draw_*` calls, and `observe=True` which not skip observe calls.

Reply:
Put this comment in the source code, for the benefit of future readers!

Reply:
Big commenting/documentation pass to come! I've learned a lot about `DataTree` and its nodes by staring at its code during this PR, and I hope to write down the most useful (to me) insights as comments while the memory of what was most confusing is still fresh.